lib/internal/util.js

```diff
@@ -187,3 +191,65 @@ exports.convertToValidSignal = function convertToValidSignal(signal) {
 
   throw new Error('Unknown signal: ' + signal);
 };
+
+const kCustomPromisifiedSymbol = Symbol('util.promisify.custom');
+const kCustomPromisifyArgsSymbol = Symbol('customPromisifyArgs');
+
+function promisify(orig) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('The first argument to promisify() must be a function');
+  }
+
+  if (orig[kCustomPromisifiedSymbol]) {
+    const fn = orig[kCustomPromisifiedSymbol];
+    Object.defineProperty(fn, kCustomPromisifiedSymbol, {
+      value: fn, enumerable: false, writable: false, configurable: true
+    });
+    return fn;
+  }
+
+  if (isAsyncFunction(orig)) {
```

Comment:
> That approach would make sense as well, but the existence of a symbol override means that your expectation already doesn't hold.

Fair enough, but that's a one-line addition, and I would say the API is still easy to understand with that change because it's still easily polyfillable.

> Similarly, if you have a callback-taking function with an overridden promisification, and you convert it to an async function, then when would you need an overridden promisification on that async function?

Note that this doesn't just apply to overridden promisifications -- it's applicable for any function that returns a Promise and also accepts a callback, where the Promise and the callback don't always resolve to the same value..

I'm not trying to argue that making a function behave like this is good design. The problem is that we're changing behavior based on whether the function was created using the `async function` syntax, which is an *implementation detail* of the function and not part of its public API. If we changed behavior based on whether the function returned a Promise, then it wouldn't be a problem (although that could be difficult to determine without calling the function beforehand).

As an example: it's common for people to write async functions, and use Babel to transpile the code to an identically-behaving function that does not use async/await syntax. This works because consumers don't care whether a function is an `async function` -- they only care that it returns a Promise and never throws synchronously. With the current `util.promisify` implementation, this would not be the case -- it would be impossible for Babel to accurately transpile async functions because we would be making decisions based on the *syntax* that the function was created with. As a result, users could end up with confusing bugs where some code works with Babel, but fails when Babel is removed.